internal/providers/github/github.go

```diff
@@ -115,3 +134,128 @@ func ParseV1Config(rawCfg json.RawMessage) (*minderv1.GitHubProviderConfig, erro
+// is rate-limited. Every cache entry has an expiration time, so the cache will eventually evict
+// the rate-limited client.
+func (c *RestClient) setAsRateLimited() {
+	if c.cache != nil {
+		c.cache.Set(c.owner, c.token, db.ProviderTypeGithub, c)
+	}
+}
+
+// waitForRateLimitReset waits for token wait limit to reset. Returns error if wait time is more
+// than MaxRateLimitWait or requests' context is cancelled.
+func (c *RestClient) waitForRateLimitReset(ctx context.Context, err error) error {
+	var rateLimitError *github.RateLimitError
+	isRateLimitErr := errors.As(err, &rateLimitError)
+
+	if isRateLimitErr {
+		return c.processPrimaryRateLimitErr(ctx, rateLimitError)
+	}
+
+	var abuseRateLimitError *github.AbuseRateLimitError
+	isAbuseRateLimitErr := errors.As(err, &abuseRateLimitError)
+
+	if isAbuseRateLimitErr {
+		return c.processAbuseRateLimitErr(ctx, abuseRateLimitError)
+	}
+
+	return nil
+}
+
+func (c *RestClient) processPrimaryRateLimitErr(ctx context.Context, err *github.RateLimitError) error {
+	logger := zerolog.Ctx(ctx)
+	rate := err.Rate
+	if rate.Remaining == 0 {
+		c.setAsRateLimited()
+
+		waitTime := DefaultRateLimitWaitTime
+		resetTime := rate.Reset.Time
+		if !resetTime.IsZero() {
+			waitTime = time.Until(resetTime)
+		}
+
+		logRateLimitError(logger, waitTime, c.owner, err.Response)
+
+		if waitTime > MaxRateLimitWait {
+			logger.Debug().Msgf("rate limit reset time: %v exceeds maximum wait time: %v", waitTime, MaxRateLimitWait)
+			return err
+		}
+
+		// Wait for the rate limit to reset
+		select {
+		case <-time.After(waitTime):
+			return nil
+		case <-ctx.Done():
+			logger.Debug().Err(ctx.Err()).Msg("context done while waiting for rate limit to reset")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *RestClient) processAbuseRateLimitErr(ctx context.Context, err *github.AbuseRateLimitError) error {
+	logger := zerolog.Ctx(ctx)
+	c.setAsRateLimited()
+
+	retryAfter := err.RetryAfter
+	waitTime := DefaultRateLimitWaitTime
+	if retryAfter != nil && *retryAfter > 0 {
+		waitTime = *retryAfter
+	}
+
+	logRateLimitError(logger, waitTime, c.owner, err.Response)
+
+	if waitTime > MaxRateLimitWait {
+		logger.Debug().Msgf("abuse rate limit wait time: %v exceeds maximum wait time: %v", waitTime, MaxRateLimitWait)
+		return err
+	}
+
+	// Wait for the rate limit to reset
+	select {
+	case <-time.After(waitTime):
+		return nil
+	case <-ctx.Done():
+		logger.Debug().Err(ctx.Err()).Msg("context done while waiting for rate limit to reset")
+		return err
+	}
+}
+
+func logRateLimitError(logger *zerolog.Logger, waitTime time.Duration, owner string, resp *http.Response) {
+	var method, path string
+	if resp != nil && resp.Request != nil {
+		method = resp.Request.Method
+		path = resp.Request.URL.Path
+	}
+
+	logger.Debug().
+		Str("owner", owner).
+		Str("method", method).
+		Str("path", path).
+		Str("wait_time", waitTime.String()).
+		Msg("rate limit exceeded")
```

Comment:
Do you think it might make sense to also log the two different rate limits (abuse vs primary) using a keyword (e.g. pass another parameter like "type" and set it one of two constant values)?

Reply:
Yes, that would be better. Done.